chainer/testing/math_function_test.py

```diff
@@ -0,0 +1,207 @@
+    return getattr(F, name, None)
+
+
+def make_data_default(dtype, shape):
+    x = numpy.random.uniform(-1, 1, shape).astype(dtype)
+    gy = numpy.random.uniform(-1, 1, shape).astype(dtype)
+    return x, gy
+
+
+def math_function_test(func, func_expected=None, label_expected=None,
+                       make_data=None):
+    """Decorator to test Chainer functions lifting mathematical numpy functions.
+
+    This decorator is for testing Chainer functions lifted from corresponding
+    mathematical numpy and cupy functions, and optionally ones composed with
+    such other Chainer functions. Forward and backward computations on CPU and
+    GPU across parameterized ``dtype`` and ``shape`` are tested.
+
+    Args:
+        func(~chainer.Function): Chainer function to be tested by
+            decorated test class.
+        func_expected: Function that is used on testing forward computation to
+            get expected values. If not given, a corresponding numpy function
+            for ``func`` is implicitly picked up from its name.
+        label_expected(string): String that is used on testing a Chainer
+            function's label to get expected one. If not given, the name of
+            ``func`` is implicitly used. If given for Chainer function that
+            does not have its label, raises ``ValueError``.
+        make_data: Function that takes ``dtype`` and ``shape`` to
+            return a tuple of input and gradient data. If not given, default
+            input and gradient are used.
+
+    ``func`` takes a Chainer function to be tested and usually it is enough.
+    ``func_expected`` is used on testing Chainer functions composed with others
+    and to give their expected values. ``label_expected`` is used to test
+    Chainer functions that override their labels. ``make_data`` is used to
+    customize input and gradient data for testing. By default, uniform
+    distribution ranged ``[-1, 1]`` is used for both.
+
+    Decorated test class tests forward and backward computation for CPU and GPU
+    across the following :func:`~chainer.testing.parameterize` ed parameters:
+
+    - dtype: ``numpy.float16``, ``numpy.float32`` and ``numpy.float64``
+    - shape: rank of zero and more
+
+    Additionally, it tests the label of Chainer function class if a Chainer
+    function has its corresponding function class. Decorator searches a Chainer
+    function class in ``chainer.functions`` module from the name of the Chainer
+    function.
+
+    .. admonition:: Example
+
+       The following code defines a test class that tests trigonometric ``sin``
+       Chainer function that takes a variable with ``dtype`` of float and
+       returns another with the same ``dtype``.
+
+       >>> import unittest
+       >>> from chainer import testing
+       >>> from chainer import functions as F
+       >>>
+       >>> @testing.math_function_test(F.sin)
+       >>> class TestSin(unittest.TestCase):
+       >>>     pass
+
```

Comment:
I tried to make it tested in `doctest` directive. I found that the following doctest tested parameterized conterparts of `TestSin` test class, but including GPU tests:

```
.. doctest::

   >>> import unittest
   >>> from chainer import testing
   >>> from chainer import functions as F
   >>> import nose
   >>>
   >>> @testing.math_function_test(F.sin)
   ... class TestSin(unittest.TestCase):
   ...     pass
   ...
   >>> nose.main(defaultTest=__name__, argv=[''], exit=False).sucess
   True
```

And the following modification, adding `"--attr='!gpu'"` to `nose.main` 's `argv` parameter, did not work, testing no tests with just returning `True`.

```
.. doctest::

   >>> import unittest
   >>> from chainer import testing
   >>> from chainer import functions as F
   >>> import nose
   >>>
   >>> @testing.math_function_test(F.sin)
   ... class TestSin(unittest.TestCase):
   ...     pass
   ...
   >>> nose.main(defaultTest=__name__, argv=['', "--attr='!gpu'"], exit=False).sucess
   True
```

I'm not sure why it does not work with refering `nose` 's document, need more time to dive into its implementation. Let me talk to you about it.

(`nose` seems to be basically designed to be used in command line, not well in scripting maybe because of repeated update. It often went not as I expected, and required me hours of study to get here.)